### PR TITLE
fix rigol dg4000 value mappings

### PR DESCRIPTION
**Pull request by konosubakonoakua**
fix value mappings

### CODE REVIEW

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py [hunk 1]
```diff
@@ -238,14 +238,14 @@ def __init__(
                 ch + "output_polarity",
                 get_cmd=output + "POL?",
                 set_cmd=output + "POL {}",
-                val_mapping={"normal": "NORM", "inverted": "INV"},
```
Comment (jenshnielsen) [inline src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: This does not look correct. The val mapping should not need to strip ctr chars like `\n` and `\r` This is usually a sign that the terminator has been configured incorrectly. By default this driver is configured to use a single `\n` as a terminator on messages. This means that the instrument driver automatically adds a `\n` to each command it sends and strips a `\n` from all messages that it reads back.  Could you check with the manual if that is correct

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: I believe that we should strip all `\n` instead of only strip the last one.

Comment (jenshnielsen) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: When you create your instrument, you can pass terminator="\n\n" to set that for your instance. But if the default for this instrument is `\n\n` we should change the default_terminator attribute on this instrument

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]:
```python
<class 'KeyError'> ("'NORMAL\n' not in val_mapping", 'getting awg_ch1_output_polarity') <traceback object at 0x000001AF28635E00>
```
Here' the error message.

Comment (jenshnielsen) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: @konosubakonoakua That makes sense. If you pass `terminator="\n\n"` when you create the instrument that error should go away.

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: > @konosubakonoakua That makes sense. If you pass `terminator="\n\n"` when you create the instrument that error should go away.

Unfortunately, pyvisa doesn't allow us to set `\n\n` as terminator.
```bash
[111](file:///C:/Users/me/Documents/VboxShared/impcas/imp_blm_test/.venv/Lib/site-packages/pyvisa/resources/messagebased.py:111) self.set_visa_attribute(
    [112](file:///C:/Users/me/Documents/VboxShared/impcas/imp_blm_test/.venv/Lib/site-packages/pyvisa/resources/messagebased.py:112)     constants.ResourceAttribute.termchar_enabled, constants.VI_TRUE
    [113](file:///C:/Users/me/Documents/VboxShared/impcas/imp_blm_test/.venv/Lib/site-packages/pyvisa/resources/messagebased.py:113) )

ValueError: ambiguous ending in termination characters
```

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: I have done some experiments, and it seems that when reading `:OUTP:POL?\n` something strange happened.
The side effect is that the Pyvisa handle seems to have lost the terminator set by qcodes instrument.

Here's the scripts:
```python
#%%
import pyvisa
from pprint import pprint
pyvisa.ResourceManager('@py')
rm = pyvisa.ResourceManager()
inst_dg4102 = rm.open_resource("TCPIP0::10.10.9.198::INSTR")
# inst_dg4102.read_termination = "\n"
# inst_dg4102.write_termination = "\n"
def query_pol_without_term():
    print(bytes(inst_dg4102.query(":OUTP:POL?\n").encode()))# type:ignore

def query_pol_with_term():
    inst_dg4102.read_termination = "\n" # type:ignore
    print(bytes(inst_dg4102.query(":OUTP:POL?\n").encode())) # type: ignore


#%%
from typing import Callable
def traceit(func:Callable):
    import inspect
    import sys

    def trace_calls(frame, event, arg):
        if event == 'call':
            co = frame.f_code
            func_name = co.co_name
            func_filename = co.co_filename
            func_lineno = frame.f_lineno
            print(f'Calling {func_name} at {func_filename}:{func_lineno}')
    sys.settrace(trace_calls)
    func()
    sys.settrace(None)

#%%
if __name__ == "__main__":
    traceit(query_pol_without_term)
    print("*"*100)
    traceit(query_pol_with_term)
```
And I get the tracing result.
```bash
python .\tracing2.py
Calling query_pol_without_term at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\tracing2.py:9
Calling query at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:624
Calling write at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:158
Calling write_raw at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:142
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling write at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\functions.py:2771
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:91
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:53
Calling _return_handler at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:184
Calling _args_to_str at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:50
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1839
Calling log at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1880
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1889
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling handle_return_value at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:229
Calling __call__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:688
Calling __new__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:1095
Calling read at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:454
Calling _read_raw at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:406
Calling ignore_warning at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:245
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling helper at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:299
Calling __init__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:104
Calling __enter__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:132
Calling ignore_warning at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:281
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1467
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling read at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\functions.py:2312
Calling create_string_buffer at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\ctypes\__init__.py:48
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:91
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:53
Calling _return_handler at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:184       
Calling _args_to_str at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:50
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1839
Calling log at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1880
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1889
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling handle_return_value at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:229
Calling __call__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:688
Calling __new__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:1095
Calling __exit__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:141
Calling ignore_warning at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:298
b'NORMAL\n\n'

****************************************************************************************************
I'm the divider.
****************************************************************************************************

Calling query_pol_with_term at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\tracing2.py:12
Calling read_termination at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:95    
Calling set_visa_attribute at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:356     
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling set_attribute at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\functions.py:2394        
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:91
Calling _return_handler at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:184       
Calling _args_to_str at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:50
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1839
Calling log at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1880
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1889
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling handle_return_value at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:229
Calling __call__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:688
Calling __new__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:1095
Calling set_visa_attribute at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:356     
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling set_attribute at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\functions.py:2394        
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:91
Calling _return_handler at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:184       
Calling _args_to_str at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:50
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1839
Calling log at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1880
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1889
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling handle_return_value at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:229
Calling __call__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:688
Calling __new__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:1095
Calling query at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:624
Calling write at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:158
Calling write_raw at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:142
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling write at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\functions.py:2771
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:91
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:53
Calling _return_handler at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:184       
Calling _args_to_str at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:50
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1839
Calling log at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1880
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1889
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling handle_return_value at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:229
Calling __call__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:688
Calling __new__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:1095
Calling read at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:454
Calling _read_raw at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\messagebased.py:406
Calling ignore_warning at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:245
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling helper at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:299
Calling __init__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:104
Calling __enter__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:132
Calling ignore_warning at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:281
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1467
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling session at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\resources\resource.py:159
Calling read at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\functions.py:2312
Calling create_string_buffer at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\ctypes\__init__.py:48
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:91
Calling from_param at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\types.py:53
Calling _return_handler at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:184
Calling _args_to_str at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:50
Calling debug at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1839
Calling log at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1880
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1889
Calling isEnabledFor at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\logging\__init__.py:1734
Calling handle_return_value at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:229
Calling __call__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:688
Calling __new__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\enum.py:1095
Calling __exit__ at C:\Users\me\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:141
Calling ignore_warning at C:\Users\me\Documents\VboxShared\impcas\imp_blm_test\.venv\Lib\site-packages\pyvisa\highlevel.py:298
b'NORMAL\n'

```

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: without `termination="\n"`, pyvisa returns "NORMAL\n\n"

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: Problem is that not all command responses have the same terminator.
Some even have no terminator at all. e.g. `:VOLT:UNIT?`
I think adding customized `get_parser` is the solution, if we must leave the `val_mapping` as it is.

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: I dont't want to spend time to dive the root cause, let's use the `get_parser` solution. I have updated the branch, please review again.

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py [hunk 2]
```diff
@@ -460,7 +465,7 @@ def __init__(
             self.add_parameter(
                 ch + "harmonic_order",
                 get_cmd=source + "HARM:ORDE?",
-                get_parser=int,
+                get_parser=float,
```
Comment (jenshnielsen) [inline src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:468]: I don't understand why this change is needed. Is it possible to set the harmonic_order to a non integer value and if so the validator below would also need to be changed. The same question applies to sweep_step_below

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:468]: okey，I will have a try. the firmware version of rigol dg4162 I used is `00.01.08`.
Actually I have already tried to update, but failed because I don't have U disk which has a capacity less than 4GB...😢. I will buy a new one.

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:468]: It's wired, right?

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py (hunk 1 above)
Comment (jenshnielsen) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: I guess we can live with the current form which just strips if required. I am ok merging that (I would still always recommend using newest firmware if possible but that is up to you)

**Comment by konosubakonoakua** [comment]
DG4000 cannot be upgraded with an arbitrary U disk.

I have consulted with Rigol Tech support, They said it's better to send my device to them.
Their engineer can do the upgrade. However, I don't have time to wait for that.
IMO, I should leave this PR open in case someone else have encountered the same issue.

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py (hunk 2 above)
Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:468]: Hi, @jenshnielsen 
I have one question: the return values of commands are defined in the [RIGOL DG4000 programming manual](https://www.rigolna.com/products/waveform-generators/dg4000/).
What I modified is based on the manual and has also been tested by myself.
It is to say that your already tested DG4000 series devices do not follow the programming manual correctly. Am I right? 🧐

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py (hunk 1 above)
Comment (jenshnielsen) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:241]: @konosubakonoakua I think your current solution with `parse_string_strip` is fine to merge so lets go ahead and do that. Can you however rename it to start with an `_` so we signal that its private and should not be used outside this driver

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py (hunk 2 above)
Comment (jenshnielsen) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:468]: @konosubakonoakua I think we should retain `get_parser=int` this means that the instrument will return `7.000000E+00` which qcodes will then convert to `7` which is the format that the instrument expects as input. This means that you can do

```
myinstr.harmonic_order.set(myinstr.harmonic_order.get()
```
E.g. the output and the input has the same format. 

There is no need for us to reveal to the end user that the instrument returns a float and takes an integer as input

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py [hunk 3]
```diff
@@ -238,13 +241,15 @@ def __init__(
                 ch + "output_polarity",
                 get_cmd=output + "POL?",
                 set_cmd=output + "POL {}",
-                val_mapping={"normal": "NORM", "inverted": "INV"},
+                get_parser=_parse_string_strip,
+                val_mapping={"normal": "NORMAL", "inverted": "INVERTED"},
```
Comment (konosubakonoakua) [inline src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: @jenshnielsen should I update the val_mapping according to the instrument manual?

**Comment by konosubakonoakua** [comment]
@jenshnielsen I have restore the parser to int, please review again.

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py (hunk 3 above)
Comment (jenshnielsen) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: I am not super familiar with this instrument but if its like most it supports both writing the short and the long form. I would say that if the long form is what is documented in the manual we should use that

**Comment by jenshnielsen** [comment]
@konosubakonoakua Should be good to go, just waiting on clarity on the last question inline

File: src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py (hunk 3 above)
Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: yes，I take the value from latest document of rigol DG4000

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: It's setting value and return value are not the same.
image

set INV
return INVERTED.

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: maybe we can wait until I confirm it again

Comment (jenshnielsen) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: Usually when docs are written such as `INVerted` it meas that `INV` is the short form and `intervted` is the long form. I think its normally case insensitive but lets use what the instrument returns. Can you check that?

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: image
image
confirmed.

Comment (konosubakonoakua) [reply src/qcodes/instrument_drivers/rigol/Rigol_DG4000.py:245]: @jenshnielsen good to go 👍

**Review by jenshnielsen — APPROVED** [review]
